### ISSUE TITLE
Pipeline archiver will archive child pipelines if `set_pipeline` is under `on_failure`, etc

### ISSUE THREAD

**Issue by clarafu** [concourse/concourse#8201]
## Summary

The automatic pipeline archiver still has some edge cases for when it will archive all the child pipelines accidentally.

One of the cases is when the parent job has the `set_pipeline` steps for its child pipelines wrapped within either an `on_failure`, `on_error`, or `on_abort`. This will result in the child pipelines getting archived every time the parent job successfully runs.

We have had to tweak the automatic pipeline archiver logic a few times (https://github.com/concourse/concourse/pull/8137 and https://github.com/concourse/concourse/pull/8137/commits/49385f20c2ef07c64d73ff8be650481b102a2a14) and knowing this bug that Evan mentioned https://github.com/concourse/concourse/pull/8179#discussion_r835127959, I'm going to think of if there is a better way to achieve the result we are looking for in a different way. 

Basically what we want is for any child pipelines to get archived when it is no longer set by its parent job. 

One thing I mentioned previously is that I'm not sure if the line of code that introduced these bugs is even necessary because we have a query that will remove unnecessary child pipelines when a build finishes

https://github.com/concourse/concourse/blob/4f3bb8fb4916cf8ff8cc918ed81580f84e418420/atc/db/build.go#L751-L762

but that only runs when the build is successful so that query will still have a problem for child pipelines that are set under `on_failure`, etc.

## Triaging info

* Concourse version: 7.7.0
* Browser (if applicable):
* Did this used to work?
